elasticdl/python/worker/allreduce_trainer.py

```diff
@@ -182,3 +184,96 @@ def export_saved_model(self, model_path):
 
     def get_model_version(self):
         return self._optimizer.iterations.numpy()
+
+
+class ElasticAllReduceController(object):
+    """The controller initializes Horovod and call the function with forward
+    and backward computation using a batch data. If Horovod raise an exception
+    about AllReduce, Allgather and Broadcast, the controller will catch the
+    exception and re-initialize Horovod. Then, it will broadcast the variables
+    and retry to call those function.
+    """
+
+    def __init__(self, master_client, master_addr):
+        if not hvd:
+            raise RuntimeError("Horovod is not installed for AllReduce")
+
+        self._rendezvous_manager = RendevousManager(master_client, master_addr)
+        self._step = 0
+        self._first_call = True
+        self._need_broadcast = True
+
+    def elastic_run(self, func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            self._init_variables_before_first_calling(func, *args, **kwargs)
+            self._init_horovod_periodically()
+            return self._try_to_call_func(func, *args, **kwargs)
+
+        return wrapper
+
+    def _init_variables_before_first_calling(self, func, *args, **kwargs):
+        if self._first_call:
+            hvd.init()
+            func(*args, **kwargs)
+            self._first_call = False
+
+    def _init_horovod_periodically(self):
+        if self._step % 20 == 0:
+            self._rendezvous_manager.init_horovod_if_needed()
+
+    def _broadcast_if_needed(self):
+        if self._rendezvous_manager.need_broadcast:
+            logger.info("Broadcast models")
+            self.broadcast()
+            self._rendezvous_manager.need_broadcast = False
+
+    def _try_to_call_func(self, func, *args, **kwargs):
```

Comment:
_try_to_call_func, func is too general. I believe you want to say `one_training_iteration`?

Reply:
Yes, the function tries to call `train_one_batch` with retries.